### PR TITLE
Feature/mapbox gl plugin

### PR DESCRIPTION
**Pull request by edbrett**
### Notes
The code is likely hard to follow and messy if you don't have experience with the MapboxGL API. I also wrote this way too fast and haven't reviewed it anyway near enough. Please be polite in your comments. I am fragile.

### Overview
A new plugin to allow agnostic use of mapbox-gl-js and provide all matching functionalities of the leaflet plugin! 🗺 💨 

Key features and achievements of the plugin:

- Layers: three new layer types:
  - `vector`: accepts any vector tiles as either the `url` prop or using the same standard `carto` config defined in the API. When using the carto config, we build the vector tiles in the same way as the `.png` tiles but instead return to the manager a Mapbox layer object.
  - `raster`: accepts a url from the prop. If a `decodeFunction` is present a custom deck.gl `TileLayer + BitmapLayer` is rendered. The decodeFunction is a valid webGL decode function.
  - `geojson`: renders any valid geojson object passed under the `data` prop. If `clusterConfig` is passed the layer is treated as a cluster layer (Mapbox has supercluster natively) and a `clusterDecode` can be used to parse the geojson from the data prop into valid point data.
- `setOpacity`: when setting opacity of Mapbox layers you must update the layers `paintProperties`. As each of these is specific to the `type` of layer you have, I have created a `buildVectorLayers` function that takes any array of vectorLayers basic config and returns the complete minimum structure for a mapbox vector layer.
- Multiple layers: vector layers require that you pass paint and layout properties for each type of layer and styles you want to render on the map. As a result you can have multiple layers for what was previously a single tile layer. As a result we have to manage these layers my assigning well name spaced ids and grouped and order them at the right times. You will see through out that we treat each layer and an array of sub layers for parsing and adding to the map.
- Default styles: any vector layer has default styles or `#f69` coloured line and fill layers to allow you to see that the layer is rendering correctly despite styles.
- `setZIndex`: Mapbox make it very hard to set z index. I am not completely sure why, but most likely because they have so much money and want to spend it on the sales team Christmas party and not on useful features for their users. As a result we have to constantly set the zIndex of the layers to make sure our data layers are below our label layers, above our basemap, and in the right order. On top of this, we have deck.gl canvas layers injected into the Mapbox canvas using the deck.gl MapboxLayer class. These don't register inside the style layer and are not accessible with `map.getStyle().layers` meaning we cannot place layers below them. As a result we have to create a background layer and group it against the canvas layer and places layers around this. Wonderful! Thanks Mapbox. I mean I might be missing something significant in their API but their terrible docs are enough of excuse for the code you find here.
- Custom decoded raster layers: many thanks to the directive of @mbarrenechea and my own dedication to making him help me, we have created two custom deck.gl layers which extend from the deck.gl `MapboxLayer` (so that we can place deck.gl layer below Mapbox style layers). This layer is a composite of a custom `TileLayer` which fetches the raster images, converts them to blobs and places them on a tile grid for the current view. These are then passed to a `BitmapLayer` which converts this layers to Bitmaps and decodes the RGBA bands using a custom shader passed from the layer parent. Finally a texture is created with the decoded pixels and placed on the map. Simple right?! A wonderful piece of hackery and over engineering for something that should be completely possible inside an enterprise mapping platform with consistent and well thought out documentation 😬 .
- Events: due to how Mapbox handles the setting of events, this plugin does not handle events for you layers in any way. These must be set in your application in the recommended way: https://docs.mapbox.com/mapbox-gl-js/api/#events.

### CODE REVIEW

File: src/plugins/plugin-mapbox-gl/index.js [hunk 1]
```diff
@@ -141,14 +141,15 @@ class PluginMapboxGL {
...
+    const layersOnMapIds = layersOnMap.map(l => l.id);
     const layersToSetIndex = layersOnMap.filter((l) => {
       const { id = {} } = l;
       const ids = mapLayers.map(ly => ly.id);
 
       return ids.includes(id);
     });
 
-    if (layerModel.decodeFunction) {
+    if (layerModel.decodeFunction && layersOnMapIds.includes(`${layerModel.id}-raster-decode`)) {
```
Comment (mbarrenechea) [inline src/plugins/plugin-mapbox-gl/index.js]: Nice

File: src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    ? layerConfig
+    : JSON.parse(replace(JSON.stringify(layerConfig), params, sqlParams));
+  let tileUrl;
+
+  const { body, url } = layerConfigParsed || {};
+  const { paint, minzoom, maxzoom } = body || {};
+
+  switch (layerModel.provider) {
+    case 'gee':
+      tileUrl = `https://api.resourcewatch.org/v1/layer/${id}/tile/gee/{z}/{x}/{y}`;
```
Comment (sorodrigo) [inline src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js]: Let’s find another solution. The library should be agnostic to the application/api

File: src/plugins/plugin-mapbox-gl/index.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+    this.map.on('style.load', () => {
+      const { getLayers } = this.options;
+      const layers = getLayers();
+
+      layers.forEach(layer => this.add(layer, layers));
+    });
+  }
+
+  method = {
+    leaflet: rasterLayer,
```
[the remark this answers is not in the document]
Comment (edbrett) [reply src/plugins/plugin-mapbox-gl/index.js]: This is to prevent a mass update in the API config for layers and also to prevent breaking changes. Ideally once we have this stable and we are happy we can remove this and release a v2.

File: src/plugins/plugin-mapbox-gl/index.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    }
+  }
+
+  /**
+   * Get provider method
+   * @param {String} provider
+   */
+  getLayerByProvider(provider, layerModel) {
+    // required to maintain current layerSpec without creating a breaking change
+    if (provider === 'leaflet' && layerModel.layerConfig.type === 'cluster') {
```
[the remark this answers is not in the document]
Comment (edbrett) [reply src/plugins/plugin-mapbox-gl/index.js]: Same here.

File: src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js (hunk 2 above)
Comment (edbrett) [reply src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js]: I cannot remember why this is here exactly, I think currently the api configs need this to be somewhere to work with rasters.

File: src/plugins/plugin-mapbox-gl/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (sorodrigo) [reply src/plugins/plugin-mapbox-gl/index.js]: You're already using it on prod. So why not dissociate what you're using in prod from what lands into layer-manager. Breaking changes can happen, in this case I would see as an option adding a new type there's no need to call it leaflet.

File: src/components/layer.js [hunk 5]
```diff
@@ -14,7 +14,11 @@ class Layer extends PureComponent {
     this.addSpecToLayerManager();
   }
 
-  componentDidUpdate() {
+  componentDidUpdate(prevProps) {
```
[the remark this answers is not in the document]
Comment (edbrett) [reply src/components/layer.js]: GeoJSON layers were not being removed when id and params were changing and I cannot for the life of my work out why. It was broken on Prod too 😄 . This fixed it and sparked joy. Agreed we need to find a proper solution.

[the remark this answers is not in the document]
Comment (sorodrigo) [reply src/components/layer.js]: I can help you out to see what's happening.

File: src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js (hunk 2 above)
Comment (sorodrigo) [reply src/plugins/plugin-mapbox-gl/raster-layer-mapbox-gl.js]: Just bringing up what we talked about offline: this should be removed and a better solution needs to be thought of before merging. What about a resource-watch service? That lives outside of the layer-manager, and a way to pass custom providers/services?
